library/rsa.c

```diff
@@ -363,11 +363,11 @@ int mbedtls_rsa_export_raw( const mbedtls_rsa_context *ctx,
 
     /* Check if key is private or public */
     is_priv =
-        mbedtls_mpi_cmp_int( &ctx->N, 0 ) != 0 &&
-        mbedtls_mpi_cmp_int( &ctx->P, 0 ) != 0 &&
-        mbedtls_mpi_cmp_int( &ctx->Q, 0 ) != 0 &&
-        mbedtls_mpi_cmp_int( &ctx->D, 0 ) != 0 &&
-        mbedtls_mpi_cmp_int( &ctx->E, 0 ) != 0;
+        ctx->N.p != NULL &&
```

Comment:
I disagree with exposing this implementation detail of `mbedtls_mpi`. If `N->n == 0` then it shouldn't make a difference whether `N->p` is null. For example, I recently [changed `mbedtls_mpi_copy`](https://github.com/ARMmbed/mbed-crypto/pull/346) not to make a difference, partly because it misled Coverity into thinking that more null checks are necessary and partly to make the code more robust against possible changes in the details of the behavior of the bignum module.

`mbedtls_rsa_complete` has treated zero (which is never a valid value for any of the fields) as meaning “to be completed” ever since it was introduced. Please keep it that way.

Reply:
If `N->n == 0`, then `mbedtls_mpi_cmp_int()` still finds it equal to 0. The only reliable way to differentiate 0 from unset is checking `N->p` at the moment.

I agree that this is not ideal and that we should not rely on implementation details, but I don't think that this PR is the best place to fix this.

Reply:
My disagreement is about treating _any_ bignum value as “unset”.

Reply:
IMO that behaviour `rsa_complete()` was a design mistake, which had unintended consequences, and since it was never documented I don't think we should feel forced to keep it.

But I take note of your opposition and will look for options we can both agree on.